flake.nix

```diff
@@ -388,6 +388,58 @@
               postUnpack = "sourceRoot=$sourceRoot/perl";
             });
 
+            passthru.python-bindings =
+              let
+                python = final.python3;
+              in python.pkgs.buildPythonPackage {
+                name = "nix";
+                format = "other";
+
+                src = self;
+                postUnpack = "sourceRoot=$sourceRoot/python";
+
+                nativeBuildInputs = [
+                  final.buildPackages.ninja
+                  (final.buildPackages.meson.override { python3 = python; })
+                  final.buildPackages.pkgconfig
+                  final.buildPackages.nix
+                ];
+
+                mesonFlags = [
+                  "-Dnix_binary=${final.lib.getBin final.nix}/bin/nix"
+                ];
+
+                strictDeps = true;
+
+                buildInputs = [
+                  final.nix
+                  final.boost
+                ];
+
+                passthru.shell = python.withPackages (p: [
+                  final.nix.python-bindings
+                ]);
+
+                installCheckPhase = ''
+                  export TEST_ROOT=$(mktemp -d)
+                  export NIX_STATE_DIR=$TEST_ROOT/var/nix
+                  export NIX_STORE_DIR=$TEST_ROOT/store
+                  export NIX_LOCALSTATE_DIR=$TEST_ROOT/var
+                  export NIX_LOG_DIR=$TEST_ROOT/var/log/nix
+
+                  export NIX_CONF_DIR=$TEST_ROOT/etc
+                  mkdir "$NIX_CONF_DIR"
+                  cat > "$NIX_CONF_DIR"/nix.conf <<EOF
+                  sandbox = false
+                  substituters =
+                  EOF
+
+                  export PYTHONPATH=$out/${python.sitePackages}
+
+                  ninja test
```

Comment:
Not sure if we should use ninja here if we're not using it elsewhere yet.  See https://github.com/NixOS/rfcs/pull/132.

Reply:
Yeah maybe, but since this is morally a separate package, just one versioned in lockstep, I don't think it is so bad.

Reply:
Turns out it's pretty much essential to not use a separate build system here, because otherwise it's really hard to iteratively work on these bindings, since the two build systems can't really call each other nicely, meaning a `nix build` is always required to actually test whether the bindings work whenever a change in Nix is done.

I'll consider a fast development cycle a requirement for this PR. I'll see if I can either make meson hackily call into autotools and depend on its built nix, or probably better, switching to autotools for the python bindings as well.

Reply:
@infinisil we currently do *not* have that for the Perl bindings, however if *both* projects were meson we would have that very nicely! https://mesonbuild.com/Subprojects.html

So I therefore consider it future work blocked on the RFC.

Reply:
> Turns out it's pretty much essential to not use a separate build system here, because otherwise it's really hard to iteratively work on these bindings, since the two build systems can't really call each other nicely

Nix proper and the bindings should be loosely coupled projects. Of course this is funny for something called a binding, but hear me out.

Nix exposes an interface
 - A change in the bindings should generally be possible to iterate on without having to change Nix.
 - A change in Nix should generally be possible to iterate on without having to change the binding.

Nix is a unit, the bindings are a unit. Each come with their own tests.

Sometimes, interfaces need to change. Such a change should be fully implemented in Nix proper first, because the fewer units you test at a time, the shorter your iteration cycles.

With these rules in mind, the workflow for an interface change will look like

1. Think of a change that affects both Nix and the binding
2. Write a unit test for the Nix part that the binding will need
3. Iterate using the unit test
4. Run the integration tests by means of `nix build`
5. Write a unit test for the binding
6. Iterate on the binding test

It'd be rare to have to iterate through 4, because that would imply that either the test (2) insufficient, or because you're in a process of validating the idea (1). The prior is just a professional growth opportunity (arrogant as that sounds; sorry), whereas the latter is unlikely to be needed for something as conceptually simple as a language binding.

The alternative is to forego componentization and always build everything and run all tests. While this might lead to a slightly simpler process for the bindings, it leads to a scalability problem for Nix, as it has to always build its bindings in order to be tested during the human part of the development cycle.

Anyway, going back to the unified build system; yes, incremental builds across the `nix` package interface would be helpful, but are not necessary for a good workflow with short cycle times.

Reply:
@roberth I think I get the idea, it would be similar to [the plugins tests](https://github.com/NixOS/nix/blob/master/tests/plugins/plugintest.cc) (which btw [have been broken multiple times](https://github.com/NixOS/nix/commits/master/tests/plugins/plugintest.cc)).

I don't think it would work very well here though, because they would have to essentially just duplicate what the bindings do, which then becomes a synchronization effort (every time you change the bindings you need to ensure the test covers the same behavior). And the only benefit I can see is a bit faster iteration times, I don't think that's worth it.

For the plugin tests this makes much more sense because that's supposed to be a stable third-party interface, meaning there's third-party code we can't test or change. The in-tree plugin test then is representative of those third-parties and (should) guard against breaking them.

Concretely for this PR, since the Nix team agreed to merge it once ready, the next best step imo is to just make sure the bindings can be iterated over quickly, _somehow_.

Reply:
> Concretely for this PR, since the Nix team agreed to merge it once ready, the next best step imo is to just make sure the bindings can be iterated over quickly, somehow.

Yeah as long as it's not a blocker for merging this, that's fine.

If this, Nix, and Hydra all use Meson, we can subproject everything and develop all 3 simultaneously which would be fantastic.